components/layout/display_list_builder.rs

```diff
@@ -1024,6 +1028,194 @@ fn convert_ellipse_size_keyword(keyword: ShapeExtent,
+        }
+
+        (BackgroundSize::Explicit { width: LengthOrPercentageOrAuto::Auto, height }, _) => {
+            let height = MaybeAuto::from_style(height, bounds_size.height)
+                                    .specified_or_default(own_size.height);
+            Size2D::new(height.scale_by(image_aspect_ratio), height)
+        }
+
+        (BackgroundSize::Explicit { width, height }, _) => {
+            Size2D::new(MaybeAuto::from_style(width, bounds_size.width)
+                                .specified_or_default(own_size.width),
+                    MaybeAuto::from_style(height, bounds_size.height)
+                                .specified_or_default(own_size.height))
+        }
+    }
+}
+
+fn tile_image_round(position: &mut Au,
+                    size: &mut Au,
+                    absolute_anchor_origin: Au,
+                    image_size: &mut Au) {
+    if *size == Au(0) || *image_size == Au(0) {
+        *position = Au(0);
+        *size =Au(0);
+        return;
+    }
+
+    let number_of_tiles = (size.to_f32_px() / image_size.to_f32_px()).round().max(1.0);
+    *image_size = *size / (number_of_tiles as i32);
+    tile_image(position, size, absolute_anchor_origin, *image_size);
+}
+
+fn tile_image_spaced(position: &mut Au,
+                     size: &mut Au,
+                     tile_spacing: &mut Au,
+                     absolute_anchor_origin: Au,
+                     image_size: Au) {
+    if *size == Au(0) || image_size == Au(0) {
+        *position = Au(0);
+        *size = Au(0);
+        *tile_spacing = Au(0);
+        return;
+    }
+
+    // Per the spec, if the space available is not enough for two images, just tile as
+    // normal but only display a single tile.
+    if image_size * 2 >= *size {
+        tile_image(position,
+                   size,
+                   absolute_anchor_origin,
+                   image_size);
+        *tile_spacing = Au(0);
+        *size = image_size;
+        return;
+    }
+
+    // Take the box size, remove room for two tiles on the edges, and then calculate how many
+    // other tiles fit in between them.
+    let size_remaining = *size - (image_size * 2);
+    let num_middle_tiles = (size_remaining.to_f32_px() / image_size.to_f32_px()).floor() as i32;
+
+    // Allocate the remaining space as padding between tiles. background-position is ignored
+    // as per the spec, so the position is just the box origin. We are also ignoring
+    // background-attachment here, which seems unspecced when combined with
+    // background-repeat: space.
+    let space_for_middle_tiles = image_size * num_middle_tiles;
+    *tile_spacing = (size_remaining - space_for_middle_tiles) / (num_middle_tiles + 1);
+}
+
+/// Tile an image
+fn tile_image(position: &mut Au,
+              size: &mut Au,
+              absolute_anchor_origin: Au,
+              image_size: Au) {
+    // Avoid division by zero below!
+    if image_size == Au(0) {
```

Comment:
I think it'd make sense to comment why the outparams are sane in this case, or why it doesn't matter..